clients/drcachesim/tests/drcachesim_unit_tests.cpp

```diff
@@ -306,9 +308,201 @@ LLC {
            num_accesses - 1);
 }
 
+// Generate a sequence of read accesses to a cache in a 2-D access pattern.
+// Loop A is the outer loop, while loop B is the inner, fastest-changing
+// loop.  The whole 2D access pattern is repeated <loop_count> times.
+// Each access is to <start_address> + <A index> * <step_size_a>
+//                                   + <B index> * <step_size_b>
+// Returns the total number of accesses performed.
+static int
+generate_2D_accesses(cache_t &cache, addr_t start_address, int step_size_a,
+                     int step_count_a, int step_size_b, int step_count_b,
+                     int loop_count = 1)
+{
+    int access_count = 0;
+    for (int i = 0; i < loop_count; ++i) {
+        memref_t ref;
+        ref.data.type = TRACE_TYPE_READ;
+        ref.data.size = 4;
+        for (int step_a = 0; step_a < step_count_a; ++step_a) {
+            for (int step_b = 0; step_b < step_count_b; ++step_b) {
+                addr_t addr = start_address + step_a * step_size_a + step_b * step_size_b;
+                ref.data.addr = addr;
+                cache.request(ref);
+                ++access_count;
+            }
+        }
+    }
+    return access_count;
+}
+
+// Convenience wrapper for a linear access pattern.
+static int
+generate_1D_accesses(cache_t &cache, addr_t start_address, int step_size, int step_count,
+                     int loop_count = 1)
+{
+    return generate_2D_accesses(cache, start_address, step_size, step_count, 0, 1,
+                                loop_count);
+}
+
+// Helper code to grab a snapshot of cache stats.
+struct CStats {
+    int_least64_t hits;
+    int_least64_t misses;
+    int_least64_t child_hits;
+};
+
+static CStats
+get_cache_stats(caching_device_stats_t &stats)
+{
+    CStats cs;
+    cs.hits = stats.get_metric(metric_name_t::HITS);
+    cs.misses = stats.get_metric(metric_name_t::MISSES);
+    cs.child_hits = stats.get_metric(metric_name_t::CHILD_HITS);
+    return cs;
+}
+
+// Creates and tests LRU caches in a range of assocativities, verifying
+// the associativity works as expected.
+void
+unit_test_cache_associativity()
+{
+    // Range of associativities to be tested.
+    constexpr int MIN_ASSOC = 1;
```

Comment:
Seems like `constexpr int` may as well always also be declared `static`

Reply:
Done.